lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.cpp

```diff
@@ -0,0 +1,235 @@
+  SmallVector<Value, 10> dispatchOperands;
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(&*functionOp.getBody().op_begin());
+  xsmm::UnaryKindAttr kind =
+      xsmm::UnaryKindAttr::get(rewriter.getContext(), unaryKind);
+  dispatchOperands.push_back(
+      getConstInt(rewriter, dyn_cast<IntegerAttr>(kind).getInt(), 64));
+
+  dispatchOperands.push_back(
+      getConstInt(rewriter, dyn_cast<DataTypeAttr>(dtype).getInt(), 64));
+
+  DenseI64ArrayAttr dims = DenseI64ArrayAttr::get(
+      rewriter.getContext(), ArrayRef<int64_t>{unaryInfo.m, unaryInfo.n,
+                                               unaryInfo.ldi, unaryInfo.ldo});
+  for (auto idx = 0; idx < dims.size(); idx++) {
+    dispatchOperands.push_back(getConstInt(rewriter, dims[idx], 64));
+  }
+  IntegerAttr unaryFlags = dyn_cast<IntegerAttr>(
+      xsmm::UnaryFlagsAttr::get(rewriter.getContext(), xsmm::UnaryFlags::NONE));
+  dispatchOperands.push_back(getConstInt(rewriter, unaryFlags.getInt(), 64));
+
+  auto dispatched = xsmm::utils::buildXsmmCall(
+      rewriter, loc, dtype, dispatchOperands,
+      IntegerType::get(rewriter.getContext(), 64), module,
+      SymbolRefAttr::get(transposeOp->getContext(), dispatchName), nullptr);
+  rewriter.setInsertionPoint(transposeOp);
+  SmallVector<Value> operandRange{
+      getConstInt(rewriter, dyn_cast<DataTypeAttr>(dtype).getInt(), 64),
+      dispatched.getResult(0), input->getOperand(0), output->getOperand(1)};
+  auto invokeCall = xsmm::utils::buildXsmmCall(
+      rewriter, loc, dtype, operandRange, TypeRange(), module,
+      SymbolRefAttr::get(transposeOp->getContext(), invokeName), output);
+  return std::make_pair(&*dispatched, &*invokeCall);
+}
+
+static std::pair<Operation *, Operation *>
+convertTransposeOp(PatternRewriter &rewriter, Operation *transposeOp,
+                   Operation *input, Operation *output, Type outputType) {
+  LLVM_DEBUG(llvm::dbgs() << "BuildTransposeOp\n");
+  Value source = input->getOperand(0);
+  VectorType outType = cast<VectorType>(outputType);
+  if (vnni::utils::isInVnniLayout(vnni::utils::VnniOperandRank::TRANSPOSE,
+                                  outType)) {
+    memref::ExpandShapeOp expandShapeOp =
+        dyn_cast<memref::ExpandShapeOp>(source.getDefiningOp());
+    source = expandShapeOp.getSrc();
+    xsmm::UnaryInfo unaryInfo;
+    unaryInfo.m = expandShapeOp.getSrcType().getShape()[0];
+    unaryInfo.n = expandShapeOp.getSrcType().getShape()[1];
+    auto stridesOnInput = mlir::utils::getStaticStrides(source);
+    unaryInfo.ldi = stridesOnInput->front();
+    auto stridesOnOutput = mlir::utils::getStaticStrides(output->getOperand(1));
+
+    // Adjust ldo based on the VNNI factor.
+    unaryInfo.ldo =
+        stridesOnOutput->front() /
+        *vnni::utils::getVnniBlockingFactor(expandShapeOp.getSrcType());
+    auto functionOp = transposeOp->getParentOfType<func::FuncOp>();
+    OpBuilder::InsertionGuard guard(rewriter);
```

Comment:
You're again setting the insertion point without reason and guards. Please do _not_ do that. This is very dangerous.

Reply:
I was trying to ensure that constants are created at the beginning of the function body. I didn't realize that the default insertion point would be the beginning of the function body. Should we not set the guard and insertion point explicitly to ensure the insertion point is set correctly or should we rely on default insertion points?

Reply:
Will delete.